### DISCUSSION TITLE
Issues with v23.9.11 , had to revoke merge

### DISCUSSION THREAD

**Discussion by helgeerbe** [hoylabs/OpenDTU-OnBattery#440] [Announcements]
Unfortunately after merging v23.9.11 I encountered several issues, so I had to revoke the upstream code.

What I observed so far:
- permanent network losses
- Crashes on MQTT HASS publish
- no live view data

At the moment, I don't know which changes from the upstream repo are causing this. So best strategy I have for the moment, to  cherry pick each commit from the upstream repo into a separate branch one after the other and test it.  This will take some time.

If there is someone out there with more time, please feel free to review the changes in commit eb1c2dbd8c81336d98a87d74777f36c0c872ee45 in the development branch.

**Comment by madmartin** [comment]
Hi,
for such situations the "git bisect" was invented.

**Reply by helgeerbe** [reply]
Thanks @madmartin wasn't aware of this command.

**Comment by helgeerbe** [comment]
I found this:

Update espressif32 from 6.3.2 to 6.4.0

```ìni
platform = espressif32@6.4.0
````

This results in crashes:

````log
12:57:22.364 > Connected to MQTT.
12:57:22.481 > Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
12:57:22.486 > 
12:57:22.486 > Core  1 register dump:
12:57:22.486 > PC      : 0x4008c069  PS      : 0x00060130  A0      : 0x8011cefb  A1      : 0x3ffd2b90  
12:57:22.498 > A2      : 0x00000000  A3      : 0xfffffffc  A4      : 0x000000ff  A5      : 0x0000ff00  
12:57:22.503 > A6      : 0x00ff0000  A7      : 0xff000000  A8      : 0x00000000  A9      : 0x3ffd2b60  
12:57:22.509 > A10     : 0x00000000  A11     : 0xffffffff  A12     : 0x00000000  A13     : 0x3ffca7a8  
12:57:22.520 > A14     : 0x00000000  A15     : 0x3ffff454  SAR     : 0x00000019  EXCCAUSE: 0x0000001c  
12:57:22.525 > EXCVADDR: 0x00000000  LBEG    : 0x4008c069  LEND    : 0x4008c079  LCOUNT  : 0xffffffff  
12:57:22.536 > 
12:57:22.536 > 
12:57:22.536 > Backtrace: 0x4008c066:0x3ffd2b90 0x4011cef8:0x3ffd2ba0 0x400ec9a9:0x3ffd2bc0 0x400e5535:0x3ffd2bf0 0x400e6709:0x3ffd2c50 0x400e68ff:0x3ffd2f60 0x400e6a21:0x3ffd2f90 0x4010b11d:0x3ffd2fb0 0x40133079:0x3ffd2fd0
12:57:23.063 > 
12:57:23.063 >   #0  0x4008c066:0x3ffd2b90 in strlen at /Users/brnomac003/.gitlab-runner/builds/qR2TxTby/0/idf/crosstool-NG/.build/xtensa-esp32-elf/src/newlib/newlib/libc/machine/xtensa/strlen.S:82
12:57:23.063 >   #1  0x4011cef8:0x3ffd2ba0 in MqttClient::publish(char const*, unsigned char, bool, char const*) at .pio/libdeps/d1_mini_esp32/espMqttClient/src/MqttClient.cpp:166
12:57:23.063 >   #2  0x400ec9a9:0x3ffd2bc0 in MqttSettingsClass::publishGeneric(String const&, String const&, bool, unsigned char) at src/MqttSettings.cpp:206
12:57:23.063 >   #3  0x400e5535:0x3ffd2bf0 in MqttHandleVedirectHassClass::publish(String const&, String const&) at src/MqttHandlVedirectHass.cpp:188 (discriminator 6)
12:57:23.063 >   #4  0x400e6709:0x3ffd2c50 in MqttHandleVedirectHassClass::publishSensor(char const*, char const*, char const*, char const*, char const*, char const*) at src/MqttHandlVedirectHass.cpp:131 (discriminator 1)
12:57:23.063 >   #5  0x400e68ff:0x3ffd2f60 in MqttHandleVedirectHassClass::publishConfig() at src/MqttHandlVedirectHass.cpp:64
12:57:23.063 >       (inlined by) MqttHandleVedirectHassClass::publishConfig() at src/MqttHandlVedirectHass.cpp:42
12:57:23.063 >   #6  0x400e6a21:0x3ffd2f90 in MqttHandleVedirectHassClass::loop() at src/MqttHandlVedirectHass.cpp:30
12:57:23.063 >   #7  0x4010b11d:0x3ffd2fb0 in loop() at src/main.cpp:222
12:57:23.063 >   #8  0x40133079:0x3ffd2fd0 in loopTask(void*) at /Users/herbe/.platformio/packages/framework-arduinoespressif32/cores/esp32/main.cpp:50
12:57:23.063 > 
12:57:23.063 > 
12:57:23.063 > 
12:57:23.063 > 
12:57:23.063 > ELF file SHA256: 5299e22061586274
````

Can anybody confirm this? I'm wondering why there are no issues with @tbnobody upstream version

**Reply by helgeerbe** [reply]
OK, without this upgrading espressif32 the merge works on my system. Will merge it to developement.  But I need some feedback from others how this update behave on their system.

**Reply by schlimmchen** [reply]
> Update espressif32 from 6.3.2 to 6.4.0

Interesting, this would have been my first guess/try if there would have been no news until this evening.

Is it always a similar backtrace? Does this happen repeatedly on your system?

Why this error does not occur in the upstream project is probably explained easily: There is no `MqttHandleVedirectHassClass` in the upstream project. AFAIK there have been no changes to `MqttSettingsClass::publishGeneric` or `MqttClient::publish`. But why would `strlen` change?

Odd. I hope I can contribute this evening.

**Reply by helgeerbe** [reply]
Thanks @schlimmchen Yes the the backtrace is still the same. What I also see, that the wifi connection becomes unreliable.

````log
17:26:33.395 > WiFi connected
17:26:33.452 > WiFi got ip: 192.168.178.87
17:26:33.453 > Network connected
17:26:33.456 > Connecting to MQTT...
17:26:35.850 > WiFi disconnected
17:26:35.851 > Try reconnecting
17:26:35.856 > Network lost connection
17:26:38.713 > WiFi disconnected
17:26:38.714 > Try reconnecting
17:26:38.718 > Network lost connection
17:26:40.471 > [ 12674][E][WiFiGeneric.cpp:1578] hostByName(): DNS Failed for home.erbehome.de
17:26:40.484 > [     12.688] DPL: waiting for valid date and time to be available
````

**Reply by schlimmchen** [reply]
Cannot confirm. :man_shrugging: At least not within the first ten minutes of runtime. Your log suggests that this happens after only 12 seconds of runtime on your board. I enabled HA integration but that did not do it either.

**Reply by schlimmchen** [reply]
Regarding the backtrace... There is multiple things very wrong that make me roll my eyes, but the one thing that might be a real candidate is this: Have a look at `MqttHandleHassClass::publishField` and `MqttHandleVedirectHassClass::publishBinarySensor`, which are part of the backtrace (there are more functions with the same issue). See how
* The `DynamicJsonDocument` is allocating a memory pool of 1024 Bytes.
* The serialization buffer is only 512 Bytes in size.
* [The docs](https://arduinojson.org/v6/api/json/serializejson/) clearly state that the overload in use only adds a zero-terminator if the buffer can hold it.
* The return value of the serialization is ignored.

So my best idea right now is that somehow the JSON increased and the serialized JSON is now larger than 512 Bytes. The buffer is then implicitly converted to a `String`, which possibly "collects" and copies multiple kB of RAM until by chance a `\0` is encountered. Then that `String` is copied once because of improper use of it as arguments to other functions, which fails since memory is scarce. So finally the payload is `NULL`, or rather `String::c_str()` returns `NULL`, which makes `strlen()` load from 0x00.

There definitely are issues here, but I am not very confident that I found the culprit.

Is this reproducible well at your end? Maybe increase those buffers (double them to match the memory pool of the DynamicJsonDocument) and see if that fixes things. (BTW: Doing so is **not** the proper way to fix the issue(s), and there are more things to do, but it might prevent the crashes you observe, so we can know that this is indeed the issue).

**Comment by helgeerbe** [comment]
Checked the logs again. It happens also in MqttHandleHassClass

````log
17:22:46.176 > Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
17:22:46.182 > 
17:22:46.182 > Core  1 register dump:
17:22:46.182 > PC      : 0x4008c069  PS      : 0x00060130  A0      : 0x8011d787  A1      : 0x3ffd2c60  
17:22:46.193 > A2      : 0x00000000  A3      : 0xfffffffc  A4      : 0x000000ff  A5      : 0x0000ff00  
17:22:46.199 > A6      : 0x00ff0000  A7      : 0xff000000  A8      : 0x00000000  A9      : 0x3ffd2c30  
17:22:46.204 > A10     : 0x00000000  A11     : 0xffffffff  A12     : 0x00000000  A13     : 0x3ffca7b8  
17:22:46.215 > A14     : 0x00000000  A15     : 0x3fffff00  SAR     : 0x0000001e  EXCCAUSE: 0x0000001c  
17:22:46.221 > EXCVADDR: 0x00000000  LBEG    : 0x4008c069  LEND    : 0x4008c079  LCOUNT  : 0xffffffff  
17:22:46.232 > 
17:22:46.232 > 
17:22:46.232 > Backtrace: 0x4008c066:0x3ffd2c60 0x4011d784:0x3ffd2c70 0x400ecb31:0x3ffd2c90 0x400e6e11:0x3ffd2cc0 0x400e86a1:0x3ffd2d20 0x400e8bea:0x3ffd3030 0x400e8cb5:0x3ffd30a0 0x4010b9b0:0x3ffd30c0 0x40133ea9:0x3ffd30e0
17:22:46.469 > 
17:22:46.469 >   #0  0x4008c066:0x3ffd2c60 in strlen at /Users/brnomac003/.gitlab-runner/builds/qR2TxTby/0/idf/crosstool-NG/.build/xtensa-esp32-elf/src/newlib/newlib/libc/machine/xtensa/strlen.S:82
17:22:46.469 >   #1  0x4011d784:0x3ffd2c70 in MqttClient::publish(char const*, unsigned char, bool, char const*) at .pio/libdeps/d1_mini_esp32/espMqttClient/src/MqttClient.cpp:166
17:22:46.469 >   #2  0x400ecb31:0x3ffd2c90 in MqttSettingsClass::publishGeneric(String const&, String const&, bool, unsigned char) at src/MqttSettings.cpp:208
17:22:46.469 >   #3  0x400e6e11:0x3ffd2cc0 in MqttHandleHassClass::publish(String const&, String const&) at src/MqttHandleHass.cpp:272 (discriminator 6)
17:22:46.469 >   #4  0x400e86a1:0x3ffd2d20 in MqttHandleHassClass::publishField(std::shared_ptr<InverterAbstract>, ChannelType_t, ChannelNum_t, byteAssign_fieldDeviceClass_t, bool) at src/MqttHandleHass.cpp:149 (discriminator 1)
17:22:46.469 >   #5  0x400e8bea:0x3ffd3030 in MqttHandleHassClass::publishConfig() at src/MqttHandleHass.cpp:75
17:22:46.469 >   #6  0x400e8cb5:0x3ffd30a0 in MqttHandleHassClass::loop() at src/MqttHandleHass.cpp:26
17:22:46.469 >   #7  0x4010b9b0:0x3ffd30c0 in loop() at src/main.cpp:220
17:22:46.469 >   #8  0x40133ea9:0x3ffd30e0 in loopTask(void*) at /Users/herbe/.platformio/packages/framework-arduinoespressif32/cores/esp32/main.cpp:50
17:22:46.469 > 
17:22:46.469 > 

````

**Reply by schlimmchen** [reply]
Interesting. What's the memory consumption?

**Reply by helgeerbe** [reply]
I was not able to login.

But my telegraf call to the prometheus api is logged says:

````log
17:30:53.832 > Calling /api/prometheus/metrics has temporarily run out of resources. Reason: "std::bad_alloc".
````

I saw one measurment in prometheus for free heap around 62 kByte free heap. Normal is around 98 kB free heap.

**Reply by schlimmchen** [reply]
I can't seem to find a difference in heap consumption between f7bd4a40d89f1391dbce472260ac2127fbf99114 (plus work in progress code from myself) compared to the same workspace with espressif@6.4.0.

Also, there is 160kB of free heap on my board. It is odd that you usually see only 98kB of free heap through prometheus. Maybe that's because the moment that value is determined that insanely huge buffer is allocated to create the prometheus response, so the heap consumption appears much larger. Otherwise I am quite confused that your heap consumption is so much larger.

**Reply by helgeerbe** [reply]
I have to invertes.

![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/59169507/2813eb90-3f82-40d4-8d50-480b1ca07fef)

But even on a nearly unconfigured test device (only network setup), I don't get 160kB

![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/59169507/0758d4dd-3798-4593-a028-09f1f34d55a3)

**Reply by schlimmchen** [reply]
Interesting. My test device has one inverter configured and still 160kB left. Are there significant differences between the ESP32 and ESP-S3? Also, mine uses 86% of the sketch, not 92% as yours does.

**Comment by schlimmchen** [comment]
FYI: Commit f7bd4a40d89f1391dbce472260ac2127fbf99114 unintentionally (?) reverts JK BMS: Support for MQTT (#432) (24018a14).

**Reply by helgeerbe** [reply]
Sorry, I will check this.

I cherry picked it. Should be back in development.

**Comment by helgeerbe** [comment]
I nailed it down, that I get different results for serializeJson()

Using branch https://github.com/helgeerbe/OpenDTU-OnBattery/tree/espressif_6.4.0 with `platform = espressif32@6.3.2` I get:

````log
20:43:55.839 > [MqttHandleHassClass::publishInverterNumber] size of buffer: 437 (437) 
20:43:55.845 > [MqttHandleHassClass::publishInverterNumber] name: "Luna", id: "limit_nonpersistent_relative"
````

and with `platform = espressif32@6.4.0`

````log
20:41:49.619 > [MqttHandleHassClass::publishInverterNumber] size of buffer: 2 (2) 
20:41:49.625 > [MqttHandleHassClass::publishInverterNumber] name: "Luna", id: "limit_nonpersistent_absolute"
````

Luna is the second Inverter. Solar ist my first one and is published always correctly.

**Reply by schlimmchen** [reply]
Interesting!

I looked at the code you pushed to the `espressif_6.4.0` branch and I am confused that the result of `measureJson` is 2 in the logs?! How can that be? `size_t` certainly did not overflow... Maybe you can print the serialized JSON to the console? Maybe write it directly to the `Serial` rather than `MessageOutput`, since there might be 2k of data in there.

You already increased the `DynamicJsonDocument`'s memory pool and adjusted the buffer accordingly. Are the crashes gone now?

**Comment by helgeerbe** [comment]
No, as soon you try to publish or print the buffer the system crashes (when it has an odd length). I will try to use Serial instead of MessageOutput

**Comment by helgeerbe** [comment]
I have no idea how to solve this. 

This was the simpliest test in the code:

````c++
void MqttHandleHassClass::publishInverterNumber(
    std::shared_ptr<InverterAbstract> inv, const char* caption, const char* icon, const char* category,
    const char* commandTopic, const char* stateTopic, const char* unitOfMeasure,
    int16_t min, int16_t max)
{
    char buffer[2048];
    size_t byte = 0;

    String serial = inv->serialString();

    String buttonId = caption;
    buttonId.replace(" ", "_");
    buttonId.toLowerCase();

    String configTopic = "number/dtu_" + serial
        + "/" + buttonId
        + "/config";

    String cmdTopic = MqttSettings.getPrefix() + serial + "/" + commandTopic;
    String statTopic = MqttSettings.getPrefix() + serial + "/" + stateTopic;

    DynamicJsonDocument root(2048);

    root["name"] = String(inv->name()) + " " + caption;
    Serial.printf("[MqttHandleHassClass::publishInverterNumber] root capacity of : %d \r\n", root.capacity());
    byte = serializeJson(root, buffer);
    MessageOutput.printf("[MqttHandleHassClass::publishInverterNumber] size of buffer(name): %d (%d) \r\n", byte, measureJson(root));
    Serial.println(buffer);
````

For my first inverter _solar_ I get:
````log
09:57:41.350 > [MqttHandleHassClass::publishInverterNumber] root capacity of : 2048 
09:57:41.356 > [MqttHandleHassClass::publishInverterNumber] size of buffer(name): 45 (45) 
09:57:41.363 > {"name":"Solar Limit NonPersistent Relative"}
````

For my second inverter _luna_ I always get:
````log
09:58:22.728 > [MqttHandleHassClass::publishInverterNumber] root capacity of : 0 
09:58:22.734 > [MqttHandleHassClass::publishInverterNumber] size of buffer(name): 2 (2) 
09:58:22.740 > {}
````

Dumping the free heap in MqttHandleHassClass::publishField shows the decrease in memory:

````log
09:58:43.838 > [MqttHandleHassClass::publishField] free heap=53512 
09:58:43.842 > [MqttHandleHassClass::publishField] free heap=53440 
09:58:43.847 > [MqttHandleHassClass::publishField] free heap=53368 
09:58:43.854 > [MqttHandleHassClass::publishField] size of buffer: 323
09:58:43.860 > [MqttHandleHassClass::publishField] free heap=52756 
09:58:43.866 > [MqttHandleHassClass::publishField] size of buffer: 333
09:58:43.872 > [MqttHandleHassClass::publishField] free heap=52128 
09:58:43.878 > [MqttHandleHassClass::publishField] size of buffer: 340
09:58:43.884 > [MqttHandleHassClass::publishField] free heap=51496 
09:58:43.890 > [MqttHandleHassClass::publishField] size of buffer: 325
09:58:43.896 > [MqttHandleHassClass::publishField] free heap=50880 
09:58:43.902 > [MqttHandleHassClass::publishField] size of buffer: 325
09:58:43.908 > [MqttHandleHassClass::publishField] free heap=50264 
09:58:43.914 > [MqttHandleHassClass::publishField] size of buffer: 317
09:58:43.920 > [MqttHandleHassClass::publishField] free heap=49660 
09:58:43.926 > [MqttHandleHassClass::publishField] size of buffer: 334
09:58:43.932 > [MqttHandleHassClass::publishField] free heap=48504 
09:58:43.937 > [MqttHandleHassClass::publishField] free heap=48432 
09:58:43.944 > [MqttHandleHassClass::publishField] size of buffer: 323
09:58:43.949 > [MqttHandleHassClass::publishField] free heap=47816 
09:58:43.956 > [MqttHandleHassClass::publishField] size of buffer: 289
09:58:43.961 > [MqttHandleHassClass::publishField] free heap=47220 
09:58:43.966 > [MqttHandleHassClass::publishField] free heap=47148 
09:58:43.973 > [MqttHandleHassClass::publishField] size of buffer: 352
09:58:43.979 > [MqttHandleHassClass::publishField] free heap=46464 
09:58:43.984 > [MqttHandleHassClass::publishField] free heap=46248 
09:58:43.990 > [MqttHandleHassClass::publishField] free heap=46028 
09:58:43.995 > [MqttHandleHassClass::publishField] free heap=45816 
09:58:44.001 > [MqttHandleHassClass::publishField] free heap=45592 
09:58:44.007 > [MqttHandleHassClass::publishField] free heap=45376 
09:58:44.011 > [MqttHandleHassClass::publishField] free heap=45296 
09:58:44.016 > [MqttHandleHassClass::publishField] free heap=45224 
09:58:44.021 > [MqttHandleHassClass::publishField] free heap=45152 
09:58:44.026 > [MqttHandleHassClass::publishField] free heap=45080 
09:58:44.031 > [MqttHandleHassClass::publishField] free heap=45008 
09:58:44.036 > [MqttHandleHassClass::publishField] free heap=44936 
09:58:44.041 > [MqttHandleHassClass::publishField] free heap=44864 
09:58:44.046 > [MqttHandleHassClass::publishField] free heap=44644 
09:58:44.051 > [MqttHandleHassClass::publishField] free heap=44564 
09:58:44.057 > [MqttHandleHassClass::publishField] free heap=44352 
09:58:44.062 > [MqttHandleHassClass::publishField] free heap=44140 
09:58:44.068 > [MqttHandleHassClass::publishField] free heap=43920 
09:58:44.073 > [MqttHandleHassClass::publishField] free heap=43704 
09:58:44.079 > [MqttHandleHassClass::publishField] free heap=43488 
09:58:44.084 > [MqttHandleHassClass::publishField] free heap=43416 
09:58:44.089 > [MqttHandleHassClass::publishField] free heap=43344 
09:58:44.093 > [MqttHandleHassClass::publishField] free heap=43272 
09:58:44.098 > [MqttHandleHassClass::publishField] free heap=43200 
09:58:44.103 > [MqttHandleHassClass::publishField] free heap=43128 
09:58:44.108 > [MqttHandleHassClass::publishField] free heap=43056 
09:58:44.113 > [MqttHandleHassClass::publishField] free heap=42984 
09:58:44.118 > [MqttHandleHassClass::publishField] free heap=42768 
09:58:44.123 > [MqttHandleHassClass::publishField] free heap=42728 
09:58:44.128 > [MqttHandleHassClass::publishField] free heap=42656 
09:58:44.133 > [MqttHandleHassClass::publishField] free heap=42576 
09:58:44.138 > [MqttHandleHassClass::publishField] free heap=42504 
09:58:44.143 > [MqttHandleHassClass::publishField] free heap=42432 
09:58:44.148 > [MqttHandleHassClass::publishField] free heap=41840 
09:58:44.152 > [MqttHandleHassClass::publishField] free heap=41768 
09:58:44.157 > [MqttHandleHassClass::publishField] free heap=41696 
09:58:44.162 > [MqttHandleHassClass::publishField] free heap=41624 
09:58:44.167 > [MqttHandleHassClass::publishField] free heap=41552 
09:58:44.174 > [MqttHandleHassClass::publishField] size of buffer: 343
09:58:44.179 > [MqttHandleHassClass::publishField] free heap=40916 
09:58:44.184 > [MqttHandleHassClass::publishField] free heap=40844 
09:58:44.189 > [MqttHandleHassClass::publishField] free heap=40772 
09:58:44.194 > [MqttHandleHassClass::publishField] free heap=40700 
````

At the moment I have no idea where to look for the memory leak.

**Comment by tbnobody** [comment]
Can you confirm this?

v6.4.0
![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/1753222/aad3814f-1181-4d05-83ba-849d6270a5a9)

v6.3.2
![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/1753222/a4ebf0f2-5325-462c-b3da-b098e3f84a9d)

If you can confirm this, I would open a issue at platform-espressif32. Even if the code is bad... the total amount of heap should not change except for maybe changes in the linker scripts?
And 34kb is not less.....

**Comment by helgeerbe** [comment]
@tbnobody 

I'm using a D1 mini board. @schlimmchen seems to have other values. If I'm using v6.4.0 in my fully configured productive board within seconds all free heap space is gone (see above).  So values below are taken from a separate D1 mini board (unconfigured except network). Until now I couldn't find out, where the leak is.

v6.3.2
![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/59169507/1a4997d7-620c-4923-94ee-5f04fe52e65b)

v6.4.0
![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/59169507/36b75755-b624-477d-8de1-1d16c59a783a)



But yes, I could confirm that 34kB of heap is missing.

**Comment by tbnobody** [comment]
The heap issue seems to be something which is fixed in arduino core 2.0.12 (which  depends on IDF 4.4.5 https://github.com/espressif/esp-idf/releases/tag/v4.4.5)

https://github.com/platformio/platform-espressif32/issues/1184#issuecomment-1707943379
